albatross.ml

```diff
@@ -431,181 +444,235 @@ module Make (S : Tcpip.Stack.V4V6) = struct
-    | Error s -> invalid_arg s
-    | Ok certificates -> (
-        raw_query t certificates cmd reply >|= function
+    match
+      key_cert ~is_ca:false ~cmd configuration.private_key "."
+        (X509.Certificate.subject configuration.certificate)
+    with
+    | Error s ->
+        Logs.warn (fun m ->
+            m "init: key_cert failed for %s: %s"
+              (Configuration.name_to_str configuration.name)
+              s);
+        let state =
+          {
+            error =
+              Some
+                (Fmt.str "Key certificate initialization failed with error: %s"
+                   s);
+            policies = Vmm_trie.empty;
+            configuration;
+          }
+        in
+        Lwt.return state
+    | Ok (key, cert) -> (
+        let certificates = `Single ([ cert ], key) in
+        raw_query stack configuration certificates cmd reply >>= function
         | Ok (_hdr, `Success (`Policies ps)) ->
-            let ps =
+            let policies =
               List.fold_left
-                (fun acc (name, p) -> fst (Vmm_trie.insert name p acc))
+                (fun trie (name, p) -> fst (Vmm_trie.insert name p trie))
                 Vmm_trie.empty ps
             in
-            t.policies <- ps;
-            t
-        | Ok w ->
-            Logs.err (fun m ->
-                m "albatross expected success policies, got reply %a"
-                  (Vmm_commands.pp_wire ~verbose:false)
-                  w);
-            t
+            let state = { error = None; policies; configuration } in
+            Lwt.return state
+        | Ok _w ->
+            Logs.warn (fun m ->
+                m "init: unexpected policy reply for %s"
+                  (Configuration.name_to_str configuration.name));
+            let state =
+              {
+                error = Some "Unexpected policy reply";
+                policies = Vmm_trie.empty;
+                configuration;
+              }
+            in
+            Lwt.return state
         | Error str ->
-            Logs.err (fun m -> m "albatross: error querying policies: %s" str);
-            t)
+            Logs.warn (fun m ->
+                m "init: raw_query failed for %s: %s"
+                  (Configuration.name_to_str configuration.name)
+                  str);
+            let state =
+              {
+                error =
+                  Some
+                    (Fmt.str "Failed to query this instance with error: %s" str);
+                policies = Vmm_trie.empty;
+                configuration;
+              }
+            in
+            Lwt.return state)
+
+  let init_all stack (configs : Configuration.t list) =
+    let open Lwt.Infix in
+    Lwt_list.fold_left_s
+      (fun acc_map (configuration : Configuration.t) ->
+        init stack configuration >>= fun state ->
+        Lwt.return (Albatross_map.add configuration.name state acc_map))
+      Albatross_map.empty configs
+    >|= fun final_map -> Ok final_map
```

Comment:
```suggestion
```

This function returns a `Albatross_map.t Lwt.t` -- no need to wrap in a result, since it cannot fail.